### PR TITLE
Symbol demangling

### LINKED ISSUE #71: Demangle symbol names
We can use [rustc-demangle](https://github.com/alexcrichton/rustc-demangle/) for demangling of Rust symbols.
Since it's more of a presentation problem, we should also save the original name, so it might look like this:

```rust
use rustc_demangle::demangle;

struct Symbol<'a> {
  orig_name: &'a str,
}

impl<'a> Symbol<'a> {
  fn get_name(&self) -> String { demangle(self.orig_name).to_string() }
  fn get_orig_name(&self) -> &str { self.orig_name }
}
```

The Rust-specific implementation should probably be confined to a Rust lang module (in case if we want to support more languages in the future).

### PR DESCRIPTION
**Pull request by Stupremee**
I'm not sure if this is the approach you wanted, but I think this should work.

Resolves #71

### CODE REVIEW

**Comment by Stupremee** [comment]
There's definitely something not working. Please dont merge

**Comment by nbaksalyar** [comment]
Thanks, this is great! I think we can also change the list of expected names in [tests/unwind_stack.rs](https://github.com/headcrab-rs/headcrab/blob/2b8dc1bf744b395b5e730228aec3be9a708f6344/tests/unwind_stack.rs#L51-L56) to verify the demangling routines work correctly. Do you want to do this as a part of this pull request?

**Comment by Stupremee** [comment]
Yes that's how I saw that my code is broken

**Comment by bjorn3** [comment]
I just remembered that `addr2line` already has demangling support. This also handles inlined functions. https://docs.rs/addr2line/0.13.0/addr2line/struct.FunctionName.html#method.demangle

Directly using `rustc-demangle` is useful for when there is no DWARF debuginfo though.

File: tests/unwind_stack.rs [hunk 1]
```diff
@@ -50,9 +50,9 @@ fn unwind_stack() -> Result<(), Box<dyn std::error::Error>> {
 
     let expected = &[
         "breakpoint",
-        "_ZN5hello4main17h",
-        "_ZN3std2rt10lang_start28_$u7b$$u7b$closure$u7d$$u7d$17h",
-        "_ZN3std2rt19lang_start_internal17h",
+        "hello::main::",
+        "std::rt::lang_start::{{closure}}::",
+        "std::rt::lang_start_internal::",
```
Comment (bjorn3) [inline tests/unwind_stack.rs:55]: I think `get_address_symbol_name` should not demangle the name. Instead there could be an extra method `get_address_demangled_name`.

**Comment by nbaksalyar** [comment]
> `addr2line` already has demangling support

Yeah, it also seems to be automatically choosing between the C++ & Rust mangling schemes depending on DWARF lang param, which is a nice feature if we want to support more langs in the future. So we can probably use addr2line as the default demangling function and resort to the direct `rustc-demangle` when there's no DWARF available?

**Comment by Stupremee** [comment]
> I just remembered that addr2line already has demangling support. 

That sounds good. Should I wait for #54 or add it here as well?

**Comment by bjorn3** [comment]
Please wait, it conflicts with #54.

**Comment by nbaksalyar** [comment]
#54 is merged now :+1:

File: src/symbol/sym.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    /// Returns the unmangled name of this symbol.
+    #[inline]
+    pub fn orig_name(&self) -> Option<&'data str> {
+        self.symbol.name()
+    }
+
+    /// Return the kind of this symbol.
+    #[inline]
+    pub fn kind(&self) -> SymbolKind {
+        self.symbol.kind()
```
Comment (nbaksalyar) [inline src/symbol/sym.rs:32]: I'd suggest adding `impl Deref<Target = object::Symbol>` to replace functions which have 1-to-1 mappings.
I.e., something like this:
```rust
impl Deref for Symbol {
    type Target = object::Symbol;

    fn deref(&self) -> &Self::Target {
        &self.symbol
    }
}
```

File: src/symbol/sym.rs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+/// A symbol table entry.
+#[derive(Clone, Debug)]
+pub struct Symbol<'data> {
+    symbol: object::Symbol<'data>,
+}
+
+impl<'data> Symbol<'data> {
+    /// Returns the demangled name if this symbol has a name.
+    #[inline]
+    pub fn name(&self) -> Option<String> {
```
Comment (bjorn3) [inline src/symbol/sym.rs:17]:
```suggestion
    pub fn demangled_name(&self) -> Option<String> {
```

Then `orig_name` is not necessary anymore.

File: src/symbol/mod.rs [hunk 4]
```diff
@@ -157,13 +159,14 @@ impl<'a> ParsedDwarf<'a> {
...
             .collect();
-        symbols.sort_by_key(|sym| sym.address());
+        symbols.sort_by_key(|sym: &Symbol| sym.address());
 
         let mut symbol_names = HashMap::new();
         for sym in &symbols {
             if let Some(name) = sym.name() {
-                symbol_names.insert(name, sym.address() as usize);
+                symbol_names.insert(name.to_string(), sym.address() as usize);
             }
```
Comment (bjorn3) [inline src/symbol/mod.rs:170]:
```suggestion
            }
            if let Some(name) = sym.demangled_name() {
                symbol_names.insert(name, sym.address() as usize);
            }
```

File: src/symbol/sym.rs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+#[derive(Clone, Debug)]
+pub struct Symbol<'data> {
+    symbol: object::Symbol<'data>,
+}
+
+impl<'data> Symbol<'data> {
+    /// Returns the demangled name if this symbol has a name.
+    #[inline]
+    pub fn demangled_name(&self) -> Option<String> {
+        // TODO: Avoid this allocation in every call. (lifetime errors)
```
Comment (bjorn3) [inline src/symbol/sym.rs:18]: You can store the demangled name in field and then return a reference.

**Comment by bjorn3** [comment]
Can you change https://github.com/headcrab-rs/headcrab/blob/25b85b7fb8129e3368ec62e0f5fc9ab89f80091a/examples/repl.rs#L230 to use the demangled name?

**Comment by nbaksalyar** [comment]
Thank you!

**Pull request merged**